images/virtualization-artifact/pkg/controller/vi/internal/source/object_ref_vd.go

```diff
@@ -0,0 +1,361 @@
+	log, _ := logger.GetDataSourceContext(ctx, objectRefDataSource)
+
+	supgen := supplements.NewGenerator(common.VIShortName, vi.Name, vi.Namespace, vi.UID)
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	switch {
+	case isDiskProvisioningFinished(*condition):
+		log.Info("Disk provisioning finished: clean up")
+
+		setPhaseConditionForFinishedImage(pvc, condition, &vi.Status.Phase, supgen)
+
+		// Protect Ready Disk and underlying PVC.
+		err = ds.diskService.Protect(ctx, vi, nil, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
+
+		return CleanUpSupplements(ctx, vi, ds)
+	case common.AnyTerminating(dv, pvc):
+		log.Info("Waiting for supplements to be terminated")
+	case dv == nil:
+		log.Info("Start import to PVC")
+
+		vi.Status.Progress = "0%"
+		vi.Status.SourceUID = util.GetPointer(vdRef.GetUID())
+
+		if err != nil {
+			setPhaseConditionToFailed(condition, &vi.Status.Phase, err)
+
+			if errors.Is(err, service.ErrInsufficientPVCSize) {
+				return false, nil
+			}
+
+			return false, err
+		}
+
+		source := &cdiv1.DataVolumeSource{
+			PVC: &cdiv1.DataVolumeSourcePVC{
+				Name:      vdRef.Status.Target.PersistentVolumeClaim,
+				Namespace: vdRef.Namespace,
+			},
+		}
+
+		sc, err := ds.diskService.GetStorageClass(ctx, &ds.storageClassForPVC)
+		if err != nil {
+			setPhaseConditionToFailed(condition, &vi.Status.Phase, err)
+			return false, err
+		}
+
+		err = ds.diskService.StartWithIgnoreWFFC(ctx, *vdRef.Spec.PersistentVolumeClaim.Size, sc.GetName(), source, vi, supgen)
+		if err != nil {
+			return false, err
+		}
+
+		vi.Status.Phase = virtv2.ImageProvisioning
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = vicondition.Provisioning
+		condition.Message = "PVC Provisioner not found: create the new one."
+
+		return true, nil
+	case pvc == nil:
+		vi.Status.Phase = virtv2.ImageProvisioning
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = vicondition.Provisioning
+		condition.Message = "PVC not found: waiting for creation."
+		return true, nil
+	case ds.diskService.IsImportDone(dv, pvc):
+		log.Info("Import has completed", "dvProgress", dv.Status.Progress, "dvPhase", dv.Status.Phase, "pvcPhase", pvc.Status.Phase)
+
+		vi.Status.Phase = virtv2.ImageReady
+		condition.Status = metav1.ConditionTrue
+		condition.Reason = vicondition.Ready
+		condition.Message = ""
+
+		var imageStatus virtv2.ImageStatusSize
+		imageStatus.Stored = vdRef.Spec.PersistentVolumeClaim.Size.String()
+		imageStatus.Unpacked = vdRef.Spec.PersistentVolumeClaim.Size.String()
+		vi.Status.Size = imageStatus
+
+		vi.Status.Format = imageformat.FormatRAW
```

Comment:
What happens if  pvc on Filesystem?

Reply:
It is planned to check in a separate task